packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,339 @@
+        })
+
+        this.bucket.on('ping', (oldContacts: DhtPeer[], newContact: DhtPeer) => this.onKBucketPing(oldContacts, newContact))
+        this.bucket.on('removed', (contact: DhtPeer) => this.onKBucketRemoved(contact))
+        this.bucket.on('added', (contact: DhtPeer) => this.onKBucketAdded(contact))
+        this.bucket.on('updated', (_oldContact: DhtPeer, _newContact: DhtPeer) => {
+            // TODO: Update contact info to the connection manager and reconnect
+        })
+        this.neighborList = new SortedContactList(selfId, this.config.maxNeighborListSize)
+        this.neighborList.on('contactRemoved', (removedContact: DhtPeer, closestContacts: DhtPeer[]) => {
+            if (this.stopped) {
+                return
+            }
+            this.emit('contactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+            this.randomPeers!.addContact(removedContact)
+        })
+        this.neighborList.on('newContact', (newContact: DhtPeer, closestContacts: DhtPeer[]) =>
+            this.emit('newContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.openInternetPeers = new SortedContactList(selfId, this.config.maxNeighborListSize / 2)
+        this.openInternetPeers.on('contactRemoved', (removedContact: DhtPeer, closestContacts: DhtPeer[]) =>
+            this.emit('openInternetContactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.openInternetPeers.on('newContact', (newContact: DhtPeer, closestContacts: DhtPeer[]) =>
+            this.emit('newOpenInternetContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        
+        this.randomPeers = new RandomContactList(selfId, this.config.maxNeighborListSize)
+        this.randomPeers.on('contactRemoved', (removedContact: DhtPeer, closestContacts: DhtPeer[]) =>
+            this.emit('randomContactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.randomPeers.on('newContact', (newContact: DhtPeer, closestContacts: DhtPeer[]) =>
+            this.emit('newRandomContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+    }
+
+    private onKBucketPing(oldContacts: DhtPeer[], newContact: DhtPeer): void {
+        if (this.stopped) {
+            return
+        }
+        const sortingList: SortedContactList<DhtPeer> = new SortedContactList(this.config.ownPeerId!, 100)
+        sortingList.addContacts(oldContacts)
+        const sortedContacts = sortingList.getAllContacts()
+        this.config.connectionManager?.weakUnlockConnection(sortedContacts[sortedContacts.length - 1].getPeerDescriptor())
+        this.bucket?.remove(sortedContacts[sortedContacts.length - 1].getPeerId().value)
```

Comment:
Do we know that `sortedContacts` can't be empty?

Reply:
yes we know, because this event handles only fires if there are too many contacts in a k-bucket and we need to choose which ones to keep and which one to discard